### PR TITLE
wrapper for metric directionality

### PR DESCRIPTION
**Pull request by shaycrk**
Adds a `Metric` decorator to allow metrics to keep track of which direction indicates better results. When using the decorator, the `optimality` parameter must be specified as either 'minimize' (that is, smaller values of the metric are better, such as for fpr) or 'maximize' (larger values are better, as in precision or recall).

The primary use case here is to allow model selection methods to determine whether how they should compare several models with respect to a given metric.

Also, I'm not sure if `optimality` is the best naming here but nothing seemed great. We could go with `greater_is_better` as in the sklearn's [make_scorer](http://scikit-learn.org/stable/modules/generated/sklearn.metrics.make_scorer.html#sklearn.metrics.make_scorer) for grid searching, but that feels a bit wordy.

### CODE REVIEW

**Comment by thcrock** [comment]
This looks good! Most of these thoughts are based on the relatively poor state of this file in particular before you got to it, but often the best times for refactoring are opportunistic like this, so here we go:

1. Knowing how this can be used I think requires a bit of knowledge about decorators, which is a good reason to call it out in a test. The form of this test I think would be to assert that everything in ModelEvaluator.available_metrics has an optimality attribute. We haven't done very well in testing the metrics yet, but this is probably a good time to start now that we're adding functionality to the metric wrappers. In the future, I am thinking about using this file as a beachhead for [Hypothesis](https://hypothesis.readthedocs.io/en/latest/) testing, as some concerns here map pretty well to that style of testing (like 'returns a float given valid input').

2. It might be time to break out the metric wrappers into their own file (`catwalk.metrics` or something). I just stuck them at the top of the evaluations module because that's really where they're used, but as they grow in size they are sticking out more here.

3. Codeclimate is complaining about these, mainly just because you touched code that already has issues. Basically, the tp/fp/tn/fn functions are super duplicated.  This is not an instance of duplication I'm super concerned about, but it's also trivial to address. What do you think of a pattern like this? And what would be a proper statistical name for that shared function?
```
def a_thing(predictions_binary, labels, prediction, label):
    results = [1 if x == prediction and y == label else 0 
             for (x, y) in zip(predictions_binary, labels)]
     return int(numpy.sum(results))

def true_positives(_, predictions_binary, labels, _):
    return a_thing(predictions_binary, labels, prediction=1, label=1)
```

**Comment by shaycrk** [comment]
Thanks, Tristan! Take a look at these additional commits, which:

1. Split out `catwalk.metrics` as a separate file
2. Ensures custom metrics include an `optimality` parameter
3. Adds a test that everything in `ModelEvaluator.available_metrics` has an `optimality` attribute which is either `minimize` or `maximize`
4. Simplifies the tp/fp/tn/fn methods using `sklearn.metrics.confusion_matrix`

However, it looks like codeclimate is still complaining about the duplication of those methods, but I'm not sure we can reduce the duplication much further. Thoughts?

Also getting an error on the Travis CI build that sqlalchemy is having trouble with numpy.int64, but I'm not seeing right away how the changes here would cause that error.

**Comment by shaycrk** [comment]
Resolved the Travis CI errors by casting the confusion_matrix output to built-in python integers rather than numpy.int64.

Also, on thinking about it more, I decided to switch from `optimality` to `greater_is_better`, since it not only reflects the `sklearn` terminology but also seems less cumbersome to simply check the boolean value `metric.greater_is_better` rather than having to check for `metric.optimality == 'maximize'`.

**Comment by thcrock** [comment]
Looks good. My take on the duplication flagging is that it's the need to int-ify the confusion matrix four times instead of having the thing that the four methods call output in the correct format.  This is less of a problem than what it was before (I believe codeclimate concurs by assigning this a lower `mass` number, though I'm not sure what that means), so I don't care as much about going further.

However, the new flag about complexity in the ModelEvaluator constructor is worth fixing, I think. Instead of putting logic directly in the constructor, that can be in a `.validate()` method. This method can be called in the constructor (and probably should), but the logic shouldn't live there.

**Comment by shaycrk** [comment]
Ah, it was actually still complaining before I added the casting to `int` -- we could split that but don't think we can escape having four functions that return calls of the same thing, at least short of doing something like making fp/fn/tp/tn a parameter of a single confusion matrix function and returning the appropriate number, but it seems like it makes more sense to me to keep them separate.

Will push a change to add `validate()` to `ModelEvaluator` in just a few.

**Comment by shaycrk** [comment]
ok @thcrock -- let me know what you think of this. resolves the complexity issue, but still complaining about duplication.
